library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java

```diff
@@ -0,0 +1,357 @@
+ *   <li>Tuple gets enqueued into Executors by calling <i>enqueueTupleForProcessing</i>.
+ *       Once can override <i>processTuple</i> method in which case, its users responsibility to enqueue tuple for
+ *       processing by calling <i>enqueueTupleForProcessing</i>.</li>
+ *   <li>When tuple gets its turn of execution, <i>processTupleAsync</i> gets called from thread other than
+ *       operator thread.</li>
+ *   <li>Based on maintainTupleOrder parameter, the results are consolidated at end of every window and respective call
+ *       to <i>handleProcessedTuple</i> will be made. This call will happen in operator thread.</li>
+ * </ol>
+ *
+ * This operator can be implemented to do async and parallel operations.
+ *
+ * Use case examples:
+ * <ul>
+ *   <li>Parallel reads from external datastore/database system.</li>
+ *   <li>Doing any operations which are long running tasks per tuple but DAG io should not be blocked.</li>
+ * </ul>
+ *
+ * @param <INPUT> input type of tuple
+ * @param <RESULT> Result of async processing of tuple.
+ * @since 3.3.0
+ */
+public abstract class AbstractAsyncProcessor<INPUT, RESULT> extends BaseOperator implements IdleTimeHandler
+{
+  private static final Logger logger = LoggerFactory.getLogger(AbstractAsyncProcessor.class);
+
+  private transient ExecutorService executor;
+  private int numProcessors = 1;
+  private boolean maintainTupleOrder = false;
+  private long processTimeoutMs = 0;
+
+  /**
+   * This Queue will hold all the tuples that are in process, has completed processing but yet to be notified for
+   * processing completion.
+   */
+  private Queue<ProcessTuple> waitingTuples = Lists.newLinkedList();
+
+  @InputPortFieldAnnotation(optional = true)
+  public final transient DefaultInputPort<INPUT> input = new DefaultInputPort<INPUT>()
+  {
+    @Override public void process(INPUT input)
+    {
+      processTuple(input);
+    }
+
+    @Override public void setup(Context.PortContext context)
+    {
+      setupInputPort(context);
+    }
+  };
+
+  /**
+   * This method is exposed for implementing class so as to take care of input port based initialization.
+   * For eg. Retrieve TUPLE_CLASS attribute from input port.
+   *
+   * @param context PortContext of input port
+   */
+  protected void setupInputPort(Context.PortContext context)
+  {
+    // Do nothing. Concrete class may choose to do override this. For eg. reading TUPLE_CLASS attribute.
+  }
+
+  /**
+   * Setup method of this operator will initialize the various types of executors possible.
+   *
+   * @param context OperatorContext
+   */
+  @Override public void setup(Context.OperatorContext context)
+  {
+    if (numProcessors == 1) {
+      this.executor = Executors.newSingleThreadExecutor();
+    } else if (numProcessors == 0) {
+      this.executor = Executors.newCachedThreadPool();
+    } else if (numProcessors > 0) {
+      this.executor = Executors.newFixedThreadPool(numProcessors);
```

Comment:
Do we need to distinguish between Executors.newSingleThreadExecutor() and Executors.newFixedThreadPool(numProcessors) with numProcessors = 1 ? Would it be functionally different ?